runtime/memprof.c

```diff
@@ -939,3 +948,29 @@ CAMLprim value caml_memprof_stop(value unit)
 
   return Val_unit;
 }
+
+/**** Interface with systhread. ****/
+
+void caml_memprof_init_th_ctx(struct caml_memprof_th_ctx* ctx) {
+  ctx->suspended = 0;
+  ctx->callback_running = 0;
+}
+
+void caml_memprof_stop_th_ctx(struct caml_memprof_th_ctx* ctx) {
+  /* Make sure that no memprof callback is being executed in this
+     thread. If so, memprof data structures may have pointers to the
+     thread's stack. */
+  if(ctx->callback_running)
+    caml_fatal_error("Thread.exit called from a memprof callback.");
```

Comment:
On the other PR @jhjourdan replied:

> Technically, I could make it an exception, but the implementation would be more complex : indeed, at this point, we have already called `caml_thread_save_runtime_state` in `st_stub.c`, so we would either have to do this check earlier or do some kind of rollback. Given that this exception is anyway unlikely to be caught, I think it is fine as is.

I prefer if `caml_fatal_error` is as a general rule avoided, and from my experience this kind of reasoning is unreliable. In fact, I do not believe it is ever true: all exceptions have a reasonable chance to be caught for cleaning up before being re-raised, so even for exceptions that do end up as an `Uncaught exception` error, having an exception instead of a fatal_error makes a difference.

Naively, I would do the check earlier around the other call to caml_invalid_argument, and raise the Invalid_argument exception too.

There is another problem, though, which is the bug by which memprof silently converts synchronous exceptions into asynchronous ones (#9267). So currently if you raise `Invalid_argument`, it might be improperly caught in one of the many places in the user's codebase where `Invalid_argument` is treated as a synchronous exception. It is subjective to say which one is worst between this (which already exists for all bugs the programmer might write in a memprof callback) or an unconditional fatal_error (in code that does not exist yet).

So, can we agree that at least _eventually_ (once the bug at #9267 is fixed), there should be an exception there? If so I am fine with whichever solution you propose as a temporary solution (fatal_error or exception).

Reply:
If you decide for an exception, I can write a test for you

Reply:
Quite frankly, I would prefer keeping the fatal error. Not so long ago, there was no mechanism to prevent this kind of misuse, and nobody bothered. With this PR, there will be one, but I consider it unlikely that this code path will be executed at all. I don't want to work to make this more complex if nobody actually cares and if the code in question will never be even executed.

Note that it is not only a question of adding code to memprof.c, but it is a question of making more complex the interface between memprof and systhread. This mean that this affects people reading both pieces of code.

Reply:
> Not so long ago, there was no mechanism to prevent this kind of misuse, and nobody bothered.

I disagree with this judgement, because over at https://github.com/ocaml/ocaml/issues/9071, I volunteered to implement precisely this myself (raising if a callback is running), but you preferred to push instead for a different solution which breaks programs (and thus could not come into action before a long time if proper deprecation practice is followed). It is also a general fallacy of software development to believe that a bug never affected anyone because nobody reported it.

Reading again the old discussion, I see we were also mentioning signal handlers. Does the same bug exists for signal handlers? If so a simple adaptation of the callback_running global would be able to prevent that (and would explain the need for a flag distinct from memprof_suspended).

By the way I was wondering whether the flag should be part of Caml_state. Is there a reason it is not?

> With this PR, there will be one, but I consider it unlikely that this code path will be executed at all. I don't want to work to make this more complex if nobody actually cares and if the code in question will never be even executed.
> 
> Note that it is not only a question of adding code to memprof.c, but it is a question of making more complex the interface between memprof and systhread. This mean that this affects people reading both pieces of code.

I am not yet convinced that it is complicated to the point of deserving a special exception to the rule. `caml_thread_exit` holds the runtime lock, is it not possible to call a memprof function there to simply check the current value of the global?

Reply:
> Reading again the old discussion, I see we were also mentioning signal handlers. Does the same bug exists for signal handlers? If so a simple adaptation of the callback_running global would be able to prevent that (and would explain the need for a flag distinct from memprof_suspended).

This problem does not affect signal handlers.

> By the way I was wondering whether the flag should be part of Caml_state. Is there a reason it is not?

No memprof global variable is part of Caml_state. The reason is that it is yet unclear what the implementation of statmemprof will look like in multicore OCaml and therefore we avoid taking any decision of this kind for now. Of course, at some point we will need to port the implementation.

> I am not yet convinced that it is complicated to the point of deserving a special exception to the rule. caml_thread_exit holds the runtime lock, is it not possible to call a memprof function there to simply check the current value of the global?

Yes, and this makes one more function in the interface between statmemprof and systhread. My point.

Reply:
For the latter I am not convinced. `caml_memprof_stop_th_ctx` has no other purpose, so it gets replaced by `caml_memprof_exit_th`. There is the same amount of functions in both cases.

Reply:
`caml_memprof_stop_th_ctx` is used in #9449.

Reply:
Thank you for the clarification. I leave my acceptance review saying that this can be revisited later unchanged, because I still believe that one needs stronger reasons to call caml_fatal_error.